src/analysis/normed_space/inner_product.lean

```diff
@@ -0,0 +1,1524 @@
+-/
+instance pi_Lp.inner_product_space {ι : Type*} [fintype ι] (f : ι → Type*)
+  [Π i, inner_product_space 𝕜 (f i)] : inner_product_space 𝕜 (pi_Lp 2 one_le_two f) :=
+{ inner := λ x y, ∑ i, inner (x i) (y i),
+  norm_sq_eq_inner :=
+  begin
+    intro x,
+    have h₁ : ∑ (i : ι), ∥x i∥ ^ (2 : ℕ) = ∑ (i : ι), ∥x i∥ ^ (2 : ℝ),
+    { apply finset.sum_congr rfl,
+      intros j hj,
+      simp [←rpow_nat_cast] },
+    have h₂ : 0 ≤ ∑ (i : ι), ∥x i∥ ^ (2 : ℝ),
+    { rw [←h₁],
+      exact finset.sum_nonneg (λ (j : ι) (hj : j ∈ finset.univ), pow_nonneg (norm_nonneg (x j)) 2) },
+    simp [norm, add_monoid_hom.map_sum, ←norm_sq_eq_inner],
+    rw [←rpow_nat_cast ((∑ (i : ι), ∥x i∥ ^ (2 : ℝ)) ^ (2 : ℝ)⁻¹) 2],
+    rw [←rpow_mul h₂],
+    norm_num [h₁],
+  end,
+  conj_sym :=
+  begin
+    intros x y,
+    unfold inner,
+    rw [←finset.sum_hom finset.univ conj],
+    apply finset.sum_congr rfl,
+    rintros z -,
+    apply inner_conj_sym,
+    apply_instance
+  end,
+  nonneg_im :=
+  begin
+    intro x,
+    unfold inner,
+    rw[←finset.sum_hom finset.univ im],
+    apply finset.sum_eq_zero,
+    rintros z -,
+    exact inner_self_nonneg_im,
+    apply_instance
+  end,
+  add_left := λ x y z,
+    show ∑ i, inner (x i + y i) (z i) = ∑ i, inner (x i) (z i) + ∑ i, inner (y i) (z i),
+    by simp only [inner_add_left, finset.sum_add_distrib],
+  smul_left := λ x y r,
+    show ∑ (i : ι), inner (r • x i) (y i) = (conj r) * ∑ i, inner (x i) (y i),
+    by simp only [finset.mul_sum, inner_smul_left]
+}
+
+/-- A field `K` satisfying `is_R_or_C` is itself a `K`-inner product space. -/
+instance is_R_or_C.inner_product_space : inner_product_space 𝕜 𝕜 :=
+{ inner := (λ x y, (conj x) * y),
+  norm_sq_eq_inner := λ x, by unfold inner; rw [mul_comm, mul_conj, of_real_re, norm_sq, norm_sq_eq_def],
+  conj_sym := λ x y, by simp [mul_comm],
+  nonneg_im := λ x, by rw[mul_im, conj_re, conj_im]; ring,
+  add_left := λ x y z, by simp [inner, add_mul],
+  smul_left := λ x y z, by simp [inner, mul_assoc] }
+
+
+/-- The standard real/complex Euclidean space, functions on a finite type. For an `n`-dimensional space
+use `euclidean_space K (fin n)`.  -/
+@[reducible, nolint unused_arguments]
+def euclidean_space (𝕜 : Type*) [nondiscrete_normed_field 𝕜] [normed_algebra ℝ 𝕜] [is_R_or_C 𝕜]
+  (n : Type*) [fintype n] : Type* := pi_Lp 2 one_le_two (λ (i : n), 𝕜)
+
+section pi_Lp
+local attribute [reducible] pi_Lp
+variables {ι : Type*} [fintype ι]
+
+instance : finite_dimensional 𝕜 (euclidean_space 𝕜 ι) := by apply_instance
+
+@[simp] lemma findim_euclidean_space :
+  finite_dimensional.findim 𝕜 (euclidean_space 𝕜 ι) = fintype.card ι := by simp
+
+lemma findim_euclidean_space_fin {n : ℕ} :
+  finite_dimensional.findim 𝕜 (euclidean_space 𝕜 (fin n)) = n := by simp
+
+end pi_Lp
+
+
+/-! ### Orthogonal projection in inner product spaces -/
+
+section orthogonal
+
+open filter
+
+include 𝕜
```

Comment:
Do you need to `include 𝕜`?  This might be a naive question, but maybe you could keep the section on orthogonal projection to convex sets in the real setting only -- then, later (say starting at 
`def submodule.orthogonal (K : submodule 𝕜 α)`
) switch back to the `is_R_or_C` setting.  (I assume there exists the result that an inner product space over `is_R_or_C` is an inner product space over `ℝ`, by `restrict_scalars`?)

Reply:
Yeah, for the theorems whose statement doesn't directly involve the inner product but only the norm, Lean can't figure out what the norm is without the include. In any case, I was planning to clean up the orthogonality section after this PR is through, it should be possible to generalize most of these statements to `is_R_or_C`.

Reply:
> Yeah, for the theorems whose statement doesn't directly involve the inner product but only the norm, Lean can't figure out what the norm is without the include. 

Hmm, I would have thought `restrict_scalars` could deal with this ...?

Reply:
I think it's a different issue from those `restrict_scalars` is meant for: The problem is that when the statement only includes norms that come from an inner product space, but not actual inner products, Lean will need to know over which field the inner product is taken at some point (for instance, if in the proof one wants to rewrite the norm in terms of the inner product). This tripped me up enough that I had to ask on Zulip actually: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Implicit.20argument.20inclusion